Build of libflatpak-docs.html is broken

Flatpak used to build `doc/reference/libflatpak-docs.html`, which is used for docs.flatpak.org, but it no longer gets created. I guess there was an issue with [this commit](https://github.com/flatpak/flatpak/commit/c12f4a5b21d873374242294fdbe3aec83f2b7066) by @alexlarsson but it's all pretty esoteric to me so I'm not sure how to fix it.

I guess it's due to `docs` not being a target:
```
0 ✓ mwleeds@toolbox ~/git/flatpak2/doc/reference (fix-libflatpak-docs-build)
↳ make libflatpak-docs.html
make: *** No rule to make target 'docs', needed by 'libflatpak-docs.html'.  Stop.
```

Removing that and specifying the target manually again causes an unsuccessful attempt at generating the HTML:
```
↳ make libflatpak-docs.html
  GEN      libflatpak-docs.html
warning: failed to load external entity "/var/home/mwleeds/git/flatpak2/doc/reference/xml/flatpak-installation.xml"
[...]
```

--- a/.github/workflows/check.yml
+++ b/.github/workflows/check.yml
@@ -176,7 +176,8 @@ jobs:
         libfuse-dev ostree libostree-dev libarchive-dev libzstd-dev libcap-dev libattr1-dev libdw-dev libelf-dev python3-pyparsing \
         libjson-glib-dev shared-mime-info desktop-file-utils libpolkit-agent-1-dev libpolkit-gobject-1-dev \
         libseccomp-dev libsoup2.4-dev libcurl4-openssl-dev libsystemd-dev libxml2-utils libgpgme11-dev gobject-introspection \
-        libgirepository1.0-dev libappstream-dev libdconf-dev clang e2fslibs-dev meson socat libxau-dev libgdk-pixbuf2.0-dev
+        libgirepository1.0-dev libappstream-dev libdconf-dev clang e2fslibs-dev meson socat libxau-dev libgdk-pixbuf2.0-dev \
+        xmlto
     - name: Check out flatpak
       uses: actions/checkout@v4
       with:
@@ -195,6 +196,14 @@ jobs:
       run: ninja -C _build
     - name: Perform CodeQL Analysis
       uses: github/codeql-action/analyze@v3
+    - name: Upload docs
+      uses: actions/upload-artifact@v4
+      with:
+        if-no-files-found: error
+        overwrite: true
+        name: docs
+        path: |
+          _build/doc/
 
   valgrind:
     name: Run tests in valgrind

--- a/doc/reference/meson.build
+++ b/doc/reference/meson.build
@@ -25,7 +25,7 @@ doc_gdbus = gnome.gdbus_codegen(
   docbook : 'dbus',
 )
 
-gnome.gtkdoc(
+libflatpak_doc = gnome.gtkdoc(
   'flatpak',
   main_xml : 'libflatpak-docs.xml',
   namespace : 'flatpak',
@@ -73,3 +73,26 @@ gnome.gtkdoc(
     '--rebuild-types',
   ],
 )
+
+if xmlto.found()
+  custom_target(
+    'libflatpak-docs.html',
+    input : [
+      '../xmlto-config.xsl',
+    ],
+    output : ['libflatpak-docs.html'],
+    depends : libflatpak_doc,
+    command : [
+      xmlto,
+      '-o', meson.current_build_dir(),
+    ] + get_option('xmlto_flags') + [
+      '--skip-validation',
+      'xhtml-nochunks',
+      '-m', '@INPUT0@',
+      fs.parent(libflatpak_doc.full_path()) / 'libflatpak-docs.xml',
+    ],
+    build_by_default : true,
+    install : true,
+    install_dir : docdir,
+  )
+endif

